### PR TITLE
Round trip 'Additional Locations' on a DiagnosticData.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Note: this currently involves disabling our optimization of only reanalyzing a method body when it is edited.  

Discussions are ongoing about what to do about this.

Update:  We only disable this optimization in the presence of diagnostics that have "AdditionalLocations".  If there are no "AdditionalLocations" present, then the optimization is performed.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
reviewers: @heejaechang @JohnHamby @srivatsn @mavasani

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by heejaechang** [comment]
@mavasani can you take a look?

:+1:

**Comment by mavasani** [comment]
LGTM. Thanks @CyrusNajmabadi.

File: src/Features/Core/Portable/Diagnostics/DiagnosticData.cs [hunk 1]
```diff
@@ -198,39 +204,96 @@ public override string ToString()
...
+        public Task<Diagnostic> ToDiagnosticAsync(Project project, CancellationToken cancellationToken)
         {
-            var location = Location.None;
-            if (tree != null)
+            return ToDiagnosticAsync(project, documentIdToTree: null, cancellationToken: cancellationToken);
+        }
+
+        private async Task<Diagnostic> ToDiagnosticAsync(Project project, ImmutableDictionary<DocumentId, SyntaxTree> documentIdToTree, CancellationToken cancellationToken)
+        {
+            return Diagnostic.Create(this.Id, this.Category, this.Message, this.Severity, this.DefaultSeverity, this.IsEnabledByDefault, this.WarningLevel, this.Title, this.Description, this.HelpLink,
```
Comment (JohnHamby) [inline src/Features/Core/Portable/Diagnostics/DiagnosticData.cs]: The formatting of this statement is hard to read. More lines with some indenting might help.

Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Diagnostics/DiagnosticData.cs]: sure!

**Comment by JohnHamby** [comment]
I sign off.

**Pull request merged**